tests/test_username_api.py

```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
-    expected = get_expected_response('twitter', user, 404)
-    actual = get_response(self.app, 'twitter', user)
-
-    message = None
-    if expected != actual:
-      message = 'The provided available username ({}) returned 200'.format(user)
-
-      user = generate_random_username(15)
-      expected = get_expected_response('twitter', user, 404)
-      actual = get_response(self.app, 'twitter', user)
-
-      if expected != actual:
-        message += ' and the random username ({}) returned 200'.format(user)
-      else:
-        message += ' but {} is still available'.format(user)
-
-    if message is not None:
-      logging.getLogger().warn(message)
-
-    assert expected == actual
-
-  @pytest.mark.parametrize('user', data['facebook']['available_usernames'])
-  def test_available_facebook_username(self, user):
-    expected = get_expected_response('facebook', user, 404)
-    actual = get_response(self.app, 'facebook', user)
-
-    # this is special to facebook checking,
-    # and 'profile' field is only useful for
-    # frontend purposes. Skipping here.
-    del actual['profile']
-
-    message = None
-    if expected != actual:
-      message = 'The provided available username ({}) returned 200'.format(user)
-
-      user = generate_random_username(25)
-      expected = get_expected_response('facebook', user, 404)
-      actual = get_response(self.app, 'facebook', user)
+  @parameterized.expand(load_test_cases('available_username'), testcase_func_name=custom_name_func)
+  def test_available_name(self, website, user):
+    expected = get_expected_response(website, user, 404)
+    actual = get_response(self.app, website, user)
 
+    if website == 'facebook':
       del actual['profile']
 
-      if expected != actual:
-        message += ' and the random username ({}) returned 200'.format(user)
-      else:
-        message += ' but {} is still available'.format(user)
-
-    if message is not None:
-      logging.getLogger().warn(message)
-
-    assert expected == actual
-
-  @pytest.mark.parametrize('user', data['openhub']['available_usernames'])
-  def test_available_openhub_username(self, user):
-    expected = get_expected_response('openhub', user, 404)
-    actual = get_response(self.app, 'openhub', user)
-
     message = None
     if expected != actual:
       message = 'The provided available username ({}) returned 200'.format(user)
 
-      user = generate_random_username(30, first_char_letter=True)
-      expected = get_expected_response('openhub', user, 404)
-      actual = get_response(self.app, 'openhub', user)
+      username_pattern = self.websites['username_patterns'][website]
+      random_length = random.randint(username_pattern['min_length'], username_pattern['max_length'])  
 
-      if expected != actual:
-        message += ' and the random username ({}) returned 200'.format(user)
+      if website == 'openhub':
+        user = generate_random_username(random_length, first_char_letter=True)
```

Comment:
I think it's a generate_random_username problem, not a testing problem. Better to modify the function to take the website as parameter, and process accordingly(not do this here)

Reply:
But this is not a very scalable thing, right? As in, we can't expect to have if elif's for tons of websites in the future, it would just make the function clunky. Having it in it's separate function to satisfy each website's own quirks is better(since they will only grow)

Reply:
But passing website name will be same since it should check what options are needed for passed website.

Reply:
if elif statements are unavoidable.

Reply:
anyway, I think your way will look better.